jinahub/rankers/SimpleRanker/simpleranker.py

```diff
@@ -1,69 +1,117 @@
+        self.matches_path = matches_path
 
     @requests(on='/search')
-    def rank(self, docs: DocumentArray, parameters: Dict, *args, **kwargs):
-        traversal_paths = parameters.get('traversal_paths', self.traversal_paths)
+    def rank(self, docs: DocumentArray, parameters: Dict, **kwargs):
+        """Aggregate the score of matched chunks to the score of their parent document.
+
+        The matches of query documents that are passed in `docs` are replaced (if they
+        exist) by documents based on the aggregated score of the parent documents of
+        matched chunks - that is, by documents containing only parent id of matched
+        chunks, and the aggregated score corresponding to that id.
 
+        :param docs: The documents for which to create aggregated matches (specifically,
+            the aggregated matches will be created for documents that are on the
+            traversal paths of documents passed in this argument).
+        :param parameters: Extra parameters that can be used to override the parameters
+            set at creation of the Executor. Valid values are `traversal_paths`
+        """
+        if docs is None:
+            return
+
+        traversal_paths = parameters.get('traversal_paths', self.traversal_paths)
         for doc in docs.traverse_flat(traversal_paths):
-            matches_of_chunks = []
-            for chunk in doc.chunks:
-                matches_of_chunks.extend(chunk.matches)
-            groups = groupby(
-                sorted(matches_of_chunks, key=lambda d: d.parent_id),
-                lambda d: d.parent_id,
-            )
-            for key, group in groups:
-                chunk_match_list = list(group)
+            parents_scores = defaultdict(list)
+            for m in DocumentArray([doc]).traverse_flat([self.matches_path]):
+                parents_scores[m.parent_id].append(m.scores[self.metric].value)
+
+            # Aggregate match scores for parent document and
+            # create doc's match based on parent document of matched chunks
+            new_matches = []
+            for match_parent_id, scores in parents_scores.items():
                 if self.ranking == 'min':
-                    chunk_match_list.sort(key=lambda m: m.scores[self.metric].value)
+                    score = min(scores)
                 elif self.ranking == 'max':
-                    chunk_match_list.sort(key=lambda m: -m.scores[self.metric].value)
-                match = chunk_match_list[0]
-                match.id = chunk_match_list[0].parent_id
-                if self.ranking in ['mean_min', 'mean_max']:
-                    scores = [el.scores[self.metric].value for el in chunk_match_list]
-                    match.scores[self.metric] = sum(scores) / len(scores)
-                doc.matches.append(match)
+                    score = max(scores)
+                elif self.ranking in ['mean_min', 'mean_max']:
+                    score = sum(scores) / len(scores)
```

Comment:
do you think in edge cases we might get divide by 0 error? e.g. docs has value while no matches at `cm` level.

Reply:
This is not possible. This stems from how `scores` are created - they are a value of `parents_scores` dict. That dict is created as a `defaultdict(list)`. So when I loop over matches, I append to this dictionary - this will create a list with scores for all `parent_id`s that appear. Conversely, for "non-existing" `parent_id`s no entry in dictionary will be created, thus there will be no empty lists to loop over.

I believe that this case (dividing by 0) should be covered by the `test_docs_no_matched_chunks` test. Let me know if you think we should test another case for this